pd/src/testnet.rs

```diff
@@ -0,0 +1,671 @@
+#
+# Options:
+#   1) "null"
+#   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
+#   3) "psql" - the indexer services backed by PostgreSQL.
+# When "kv" or "psql" is chosen "tx.height" and "tx.hash" will always be indexed.
+indexer = ["kv"]
+
+# The PostgreSQL connection configuration, the connection format:
+#   postgresql://<user>:<password>@<host>:<port>/<db>?<opts>
+psql-conn = ""
+
+#######################################################
+###       Instrumentation Configuration Options     ###
+#######################################################
+[instrumentation]
+
+# When true, Prometheus metrics are served under /metrics on
+# PrometheusListenAddr.
+# Check out the documentation for the list of available metrics.
+prometheus = true
+
+# Address to listen for Prometheus collector(s) connections
+prometheus-listen-addr = ":26660"
+
+# Maximum number of simultaneous connections.
+# If you want to accept a larger number than the default, make sure
+# you increase your OS limits.
+# 0 - unlimited.
+max-open-connections = 3
+
+# Instrumentation namespace
+namespace = "tendermint"
+"#,
+        node_name
+    )
+}
+
+/// Represents initial allocations to the testnet.
+#[derive(Debug, Deserialize)]
+pub struct TestnetAllocation {
+    #[serde(deserialize_with = "string_u64")]
+    pub amount: u64,
+    pub denom: String,
+    pub address: String,
+}
+
+/// Represents a funding stream within a testnet configuration file.
+#[derive(Debug, Deserialize)]
+pub struct TestnetFundingStream {
+    pub rate_bps: u16,
+    pub address: String,
+}
+
+/// Represents testnet validators in configuration files.
+#[derive(Debug, Deserialize)]
+pub struct TestnetValidator {
+    pub name: String,
+    pub website: String,
+    pub description: String,
+    pub funding_streams: Vec<TestnetFundingStream>,
+    pub sequence_number: u32,
+    pub voting_power: u32,
+}
+
+impl From<&TestnetAllocation> for genesis::Allocation {
+    fn from(a: &TestnetAllocation) -> genesis::Allocation {
+        genesis::Allocation {
+            amount: a.amount,
+            denom: a.denom.clone(),
+            address: Address::from_str(&a.address)
+                .expect("Invalid address format in allocations.csv"),
+        }
+    }
+}
+
+#[derive(Deserialize)]
+pub struct TendermintNodeKey {
+    pub id: String,
+    pub priv_key: TendermintPrivKey,
+}
+
+#[derive(Deserialize)]
+pub struct TendermintPrivKey {
+    #[serde(rename(serialize = "type"))]
+    pub key_type: String,
+    pub value: PrivateKey,
+}
+
+// Easiest to hardcode since we never change these.
+pub fn get_validator_state() -> String {
+    r#"{
+    "height": "0",
+    "round": 0,
+    "step": 0
+}
+"#
+    .to_string()
+}
+
+/// Expand tildes in a path.
+/// Modified from https://stackoverflow.com/a/68233480
+pub fn canonicalize_path(input: &str) -> PathBuf {
```

Comment:
Will this work on non-*nix operating systems?

Reply:
I think it should. I can give it a test on my Windows machine later